servicetalk-tcp-netty-internal/src/main/java/io/servicetalk/tcp/netty/internal/EarlyConnectionAcceptorHandler.java

```diff
@@ -0,0 +1,135 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.tcp.netty.internal;
+
+import io.servicetalk.concurrent.internal.QueueFullException;
+
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+
+import java.util.ArrayDeque;
+import java.util.NoSuchElementException;
+import java.util.Queue;
+import javax.annotation.Nullable;
+
+/**
+ * Netty handler which queues pipeline events until {@link #releaseEvents()} is called or the handler is removed.
+ */
+final class EarlyConnectionAcceptorHandler extends ChannelInboundHandlerAdapter {
+
+    /**
+     * Starting with small number of elements since this queue is unlikely to contain lots of elements
+     * (optimizing for memory usage).
+     */
+    private final Queue<Runnable> queuedEvents = new ArrayDeque<>(4);
+    private boolean queueingEnabled = true;
+    @Nullable
+    private ChannelHandlerContext channelHandlerContext;
+
+    /**
+     * Releases all queued events and disables further queueing (all further events will be propagated immediately).
+     */
+    void releaseEvents() {
+        assert channelHandlerContext != null;
+        assert channelHandlerContext.channel().eventLoop().inEventLoop();
+
+        if (queueingEnabled) {
+            queueingEnabled = false;
+
+            while (true) {
+                Runnable event = queuedEvents.poll();
+                if (event == null) {
+                    break;
+                }
+                event.run();
+            }
+
+            try {
+                channelHandlerContext.pipeline().remove(this);
+            } catch (NoSuchElementException ex) {
+                // Ignore - the handler has been already removed and called from handlerRemoved
+            }
+        }
+    }
+
+    private void enqueueOrRunEvent(final Runnable event) {
+        if (queueingEnabled) {
+            if (!queuedEvents.offer(event)) {
+                throw new QueueFullException(getClass().getName() + "#queuedEvents");
+            }
+        } else {
+            event.run();
+        }
+    }
+
+    @Override
+    public void handlerAdded(final ChannelHandlerContext ctx) throws Exception {
+        channelHandlerContext = ctx;
+        super.handlerAdded(ctx);
+    }
+
+    @Override
+    public void handlerRemoved(final ChannelHandlerContext ctx) throws Exception {
+        releaseEvents();
+        super.handlerRemoved(ctx);
+    }
+
+    @Override
+    public void channelRead(final ChannelHandlerContext ctx, final Object msg) {
+        enqueueOrRunEvent(() -> ctx.fireChannelRead(msg));
+        releaseEvents();
```

Comment:
Thanks a lot for pointing me to this method offline. I finally realized how this works, why SSL case works as expected and found out that HTTP/2 plaintext fails when run repeatedly because h2 uses `AUTO_READ`. As the result, we may receive a `channelRead` event automatically, while the early acceptor is still running on an offloaded thread.

After thinking about it more, I realized that there is no need for HTTP/2 to set AUTO_READ proactively on the Bootstrap. 
Its [H2ParentConnectionContext](https://github.com/apple/servicetalk/blob/e5f0853adec1c8e8a0f3acba46453904fd349065/servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/H2ParentConnectionContext.java#L210-L213) already has a logic to override `AUTO_READ` when h2 pipeline initialized. We have to do it this way for ALPN use-case, and we can stick to this approach for plaintext h2-prior-knowledge use-case as well without any impact. I will elaborate more when we meet.

After adding the suggested tests you will see HTTP/2 plaintext failing. This patch will solve the issue:

```patch
Subject: [PATCH] Patch
---
Index: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DeferredServerChannelBinder.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DeferredServerChannelBinder.java b/servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DeferredServerChannelBinder.java
--- a/servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DeferredServerChannelBinder.java	(revision 710e6c7b6736b179eb607f2604b356fc3984c648)
+++ b/servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DeferredServerChannelBinder.java	(date 1674109872536)
@@ -68,9 +68,7 @@
                 (channel, connectionObserver) -> alpnInitChannel(listenAddress, channel, config, executionContext,
                         service, drainRequestPayloadBody, connectionObserver);
 
-        // We disable auto read by default so we can handle stuff in the ConnectionFilter before we accept any content.
-        // In case ALPN negotiates h2, h2 connection MUST enable auto read for its Channel.
-        return TcpServerBinder.bind(listenAddress, tcpConfig, false, executionContext, connectionAcceptor, channelInit,
+        return TcpServerBinder.bind(listenAddress, tcpConfig, executionContext, connectionAcceptor, channelInit,
                 serverConnection -> {
                     // Start processing requests on http/1.1 connection:
                     if (serverConnection instanceof NettyHttpServerConnection) {
Index: servicetalk-http-netty/src/test/java/io/servicetalk/http/netty/HttpRequestEncoderTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/servicetalk-http-netty/src/test/java/io/servicetalk/http/netty/HttpRequestEncoderTest.java b/servicetalk-http-netty/src/test/java/io/servicetalk/http/netty/HttpRequestEncoderTest.java
--- a/servicetalk-http-netty/src/test/java/io/servicetalk/http/netty/HttpRequestEncoderTest.java	(revision 710e6c7b6736b179eb607f2604b356fc3984c648)
+++ b/servicetalk-http-netty/src/test/java/io/servicetalk/http/netty/HttpRequestEncoderTest.java	(date 1674109872572)
@@ -436,7 +436,7 @@
 
             ReadOnlyTcpServerConfig sConfig = new TcpServerConfig().asReadOnly();
             ServerContext serverContext = resources.prepend(
-                    TcpServerBinder.bind(localAddress(0), sConfig, false,
+                    TcpServerBinder.bind(localAddress(0), sConfig,
                             SEC, null,
                             (channel, observer) -> DefaultNettyConnection.initChannel(channel, SEC.bufferAllocator(),
                                     SEC.executor(), SEC.ioExecutor(),
Index: servicetalk-http-netty/src/test/java/io/servicetalk/http/netty/FlushStrategyOnServerTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/servicetalk-http-netty/src/test/java/io/servicetalk/http/netty/FlushStrategyOnServerTest.java b/servicetalk-http-netty/src/test/java/io/servicetalk/http/netty/FlushStrategyOnServerTest.java
--- a/servicetalk-http-netty/src/test/java/io/servicetalk/http/netty/FlushStrategyOnServerTest.java	(revision 710e6c7b6736b179eb607f2604b356fc3984c648)
+++ b/servicetalk-http-netty/src/test/java/io/servicetalk/http/netty/FlushStrategyOnServerTest.java	(date 1674109872563)
@@ -30,7 +30,6 @@
 import io.servicetalk.tcp.netty.internal.TcpServerBinder;
 import io.servicetalk.tcp.netty.internal.TcpServerChannelInitializer;
 import io.servicetalk.tcp.netty.internal.TcpServerConfig;
-import io.servicetalk.transport.api.ConnectionObserver;
 import io.servicetalk.transport.api.ServerContext;
 import io.servicetalk.transport.netty.internal.ExecutionContextExtension;
 
@@ -117,15 +116,14 @@
         final ReadOnlyTcpServerConfig tcpReadOnly = new TcpServerConfig().asReadOnly();
 
         try {
-            serverContext = TcpServerBinder.bind(localAddress(0), tcpReadOnly, true,
+            serverContext = TcpServerBinder.bind(localAddress(0), tcpReadOnly,
                     httpExecutionContext, null,
                     (channel, observer) -> {
-                        final ConnectionObserver connectionObserver = config.tcpConfig().transportObserver()
-                                .onNewConnection(channel.localAddress(), channel.remoteAddress());
+                        channel.config().setAutoRead(true);
                         return initChannel(channel, httpExecutionContext, config,
-                                new TcpServerChannelInitializer(tcpReadOnly, connectionObserver)
+                                new TcpServerChannelInitializer(tcpReadOnly, observer)
                                         .andThen(channel1 -> channel1.pipeline().addLast(interceptor)), service,
-                                true, connectionObserver);
+                                true, observer);
                     },
                     connection -> connection.process(true), null, null)
                     .map(delegate -> new NettyHttpServerContext(delegate, service, httpExecutionContext))
Index: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/H2ServerParentConnectionContext.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/H2ServerParentConnectionContext.java b/servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/H2ServerParentConnectionContext.java
--- a/servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/H2ServerParentConnectionContext.java	(revision 710e6c7b6736b179eb607f2604b356fc3984c648)
+++ b/servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/H2ServerParentConnectionContext.java	(date 1674109872543)
@@ -96,8 +96,7 @@
             return failed(newH2ConfigException());
         }
         final ReadOnlyTcpServerConfig tcpServerConfig = config.tcpConfig();
-        // Auto read is required for h2
-        return TcpServerBinder.bind(listenAddress, tcpServerConfig, true, executionContext, connectionAcceptor,
+        return TcpServerBinder.bind(listenAddress, tcpServerConfig, executionContext, connectionAcceptor,
                 (channel, connectionObserver) -> initChannel(listenAddress, channel, executionContext, config,
                         new TcpServerChannelInitializer(tcpServerConfig, connectionObserver), service,
                         drainRequestPayloadBody, connectionObserver),
Index: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/NettyHttpServer.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/NettyHttpServer.java b/servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/NettyHttpServer.java
--- a/servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/NettyHttpServer.java	(revision 710e6c7b6736b179eb607f2604b356fc3984c648)
+++ b/servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/NettyHttpServer.java	(date 1674109872553)
@@ -127,8 +127,7 @@
         }
         // This state is read only, so safe to keep a copy across Subscribers
         final ReadOnlyTcpServerConfig tcpServerConfig = config.tcpConfig();
-        // We disable auto read so that we can handle stuff in the ConnectionFilter before we accept any content.
-        return TcpServerBinder.bind(address, tcpServerConfig, false, executionContext, connectionAcceptor,
+        return TcpServerBinder.bind(address, tcpServerConfig, executionContext, connectionAcceptor,
                 (channel, connectionObserver) -> initChannel(channel, executionContext, config,
                         new TcpServerChannelInitializer(tcpServerConfig, connectionObserver), service,
                         drainRequestPayloadBody, connectionObserver),
Index: servicetalk-tcp-netty-internal/src/main/java/io/servicetalk/tcp/netty/internal/TcpServerChannelInitializer.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/servicetalk-tcp-netty-internal/src/main/java/io/servicetalk/tcp/netty/internal/TcpServerChannelInitializer.java b/servicetalk-tcp-netty-internal/src/main/java/io/servicetalk/tcp/netty/internal/TcpServerChannelInitializer.java
--- a/servicetalk-tcp-netty-internal/src/main/java/io/servicetalk/tcp/netty/internal/TcpServerChannelInitializer.java	(revision 710e6c7b6736b179eb607f2604b356fc3984c648)
+++ b/servicetalk-tcp-netty-internal/src/main/java/io/servicetalk/tcp/netty/internal/TcpServerChannelInitializer.java	(date 1674109872597)
@@ -32,7 +32,7 @@
 /**
  * {@link ChannelInitializer} for TCP.
  */
-public class TcpServerChannelInitializer implements ChannelInitializer {
+public class TcpServerChannelInitializer implements ChannelInitializer {    // FIXME: 0.43 - make this class final
 
     private final ChannelInitializer delegate;
 
Index: servicetalk-tcp-netty-internal/src/main/java/io/servicetalk/tcp/netty/internal/TcpServerBinder.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/servicetalk-tcp-netty-internal/src/main/java/io/servicetalk/tcp/netty/internal/TcpServerBinder.java b/servicetalk-tcp-netty-internal/src/main/java/io/servicetalk/tcp/netty/internal/TcpServerBinder.java
--- a/servicetalk-tcp-netty-internal/src/main/java/io/servicetalk/tcp/netty/internal/TcpServerBinder.java	(revision 710e6c7b6736b179eb607f2604b356fc3984c648)
+++ b/servicetalk-tcp-netty-internal/src/main/java/io/servicetalk/tcp/netty/internal/TcpServerBinder.java	(date 1674109893157)
@@ -101,7 +101,7 @@
             @Nullable final InfluencerConnectionAcceptor connectionAcceptor,
             final BiFunction<Channel, ConnectionObserver, Single<CC>> connectionFunction,
             final Consumer<CC> connectionConsumer) {
-        return bind(listenAddress, config, autoRead, executionContext, connectionAcceptor, connectionFunction,
+        return bind(listenAddress, config, executionContext, connectionAcceptor, connectionFunction,
                 connectionConsumer, null, null);
     }
 
@@ -111,7 +111,6 @@
      *
      * @param listenAddress The address to bind to.
      * @param config The {@link ReadOnlyTcpServerConfig} to use for the bind socket and newly accepted sockets.
-     * @param autoRead if {@code true} auto read will be enabled for new {@link Channel}s.
      * @param executionContext The {@link ExecutionContext} to use for the bind socket.
      * @param connectionAcceptor The {@link ConnectionAcceptor} used to filter newly accepted sockets.
      * @param connectionFunction Used to create a new {@link NettyConnection} from a {@link Channel}.
@@ -124,7 +123,7 @@
      * listening on the {@code listenAddress}.
      */
     public static <CC extends ConnectionContext> Single<ServerContext> bind(SocketAddress listenAddress,
-            final ReadOnlyTcpServerConfig config, final boolean autoRead, final ExecutionContext<?> executionContext,
+            final ReadOnlyTcpServerConfig config, final ExecutionContext<?> executionContext,
             @Nullable final InfluencerConnectionAcceptor connectionAcceptor,
             final BiFunction<Channel, ConnectionObserver, Single<CC>> connectionFunction,
             final Consumer<CC> connectionConsumer,
@@ -135,7 +134,7 @@
         listenAddress = toNettyAddress(listenAddress);
         EventLoopAwareNettyIoExecutor nettyIoExecutor = toEventLoopAwareNettyIoExecutor(executionContext.ioExecutor());
         ServerBootstrap bs = new ServerBootstrap();
-        configure(config, autoRead, bs, nettyIoExecutor.eventLoopGroup(), listenAddress.getClass());
+        configure(config, bs, nettyIoExecutor.eventLoopGroup(), listenAddress.getClass());
 
         ChannelSet channelSet = new ChannelSet(
                 executionContext.executionStrategy().isCloseOffloaded() ? executionContext.executor() : immediate());
@@ -315,7 +314,7 @@
         return connection;
     }
 
-    private static void configure(ReadOnlyTcpServerConfig config, boolean autoRead, ServerBootstrap bs,
+    private static void configure(ReadOnlyTcpServerConfig config, ServerBootstrap bs,
                                   @Nullable EventLoopGroup eventLoopGroup,
                                   Class<? extends SocketAddress> bindAddressClass) {
         if (eventLoopGroup == null) {
@@ -335,7 +334,10 @@
             bs.option(option, opt.getValue());
         }
 
-        bs.childOption(ChannelOption.AUTO_READ, autoRead);
+        // We disable AUTO_READ by default to prevent reading any data before the decision to accept the connection is
+        // made and to allow managing back-pressure after that. Protocols that require AUTO_READ have to enable it at
+        // their handler.
+        bs.childOption(ChannelOption.AUTO_READ, false);
 
         // Set the correct ByteBufAllocator based on our BufferAllocator to minimize memory copies.
         ByteBufAllocator byteBufAllocator = POOLED_ALLOCATOR;
Index: servicetalk-tcp-netty-internal/src/testFixtures/java/io/servicetalk/tcp/netty/internal/TcpServer.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/servicetalk-tcp-netty-internal/src/testFixtures/java/io/servicetalk/tcp/netty/internal/TcpServer.java b/servicetalk-tcp-netty-internal/src/testFixtures/java/io/servicetalk/tcp/netty/internal/TcpServer.java
--- a/servicetalk-tcp-netty-internal/src/testFixtures/java/io/servicetalk/tcp/netty/internal/TcpServer.java	(revision 710e6c7b6736b179eb607f2604b356fc3984c648)
+++ b/servicetalk-tcp-netty-internal/src/testFixtures/java/io/servicetalk/tcp/netty/internal/TcpServer.java	(date 1674109872603)
@@ -82,7 +82,7 @@
                               Function<NettyConnection<Buffer, Buffer>, Completable> service,
                               ExecutionStrategy executionStrategy)
             throws ExecutionException, InterruptedException {
-        return TcpServerBinder.bind(localAddress(port), config, false, executionContext, connectionAcceptor,
+        return TcpServerBinder.bind(localAddress(port), config, executionContext, connectionAcceptor,
                 (channel, connectionObserver) -> DefaultNettyConnection.<Buffer, Buffer>initChannel(channel,
                         executionContext.bufferAllocator(), executionContext.executor(), executionContext.ioExecutor(),
                         UNSUPPORTED_PROTOCOL_CLOSE_HANDLER, config.flushStrategy(), config.idleTimeoutMs(),
Index: servicetalk-tcp-netty-internal/src/main/java/io/servicetalk/tcp/netty/internal/TcpClientChannelInitializer.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/servicetalk-tcp-netty-internal/src/main/java/io/servicetalk/tcp/netty/internal/TcpClientChannelInitializer.java b/servicetalk-tcp-netty-internal/src/main/java/io/servicetalk/tcp/netty/internal/TcpClientChannelInitializer.java
--- a/servicetalk-tcp-netty-internal/src/main/java/io/servicetalk/tcp/netty/internal/TcpClientChannelInitializer.java	(revision 710e6c7b6736b179eb607f2604b356fc3984c648)
+++ b/servicetalk-tcp-netty-internal/src/main/java/io/servicetalk/tcp/netty/internal/TcpClientChannelInitializer.java	(date 1674109872578)
@@ -33,7 +33,7 @@
 /**
  * {@link ChannelInitializer} for TCP client.
  */
-public class TcpClientChannelInitializer implements ChannelInitializer {
+public class TcpClientChannelInitializer implements ChannelInitializer {    // FIXME: 0.43 - make this class final
 
     private final ChannelInitializer delegate;
 

```

Reply:
^ I added 2 comments for 0.43. It's not expected that these classes are not final. Unlikely we will break anyone (they are internal), just trying to play extra safe game for 0.42.x